File: 20221103_Thursday/20221103.go
// Leetcode: 2114. Maximum Number of Words Found in Sentences.
// https://leetcode.com/problems/maximum-number-of-words-found-in-sentences/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 2 ms, faster than 92.23% of Go online submissions for Maximum Number of Words Found in Sentences.
// Memory Usage: 4.2 MB, less than 10.68% of Go online submissions for Maximum Number of Words Found in Sentences.
// 2022.11.04 Friday.
import "strings"
func mostWordsFound(sentences []string) int {
  max := 0
  for _, sentence := range sentences {
    size := len(strings.Split(sentence," "))
    if max < size {
      max = size
    }
  }
  return max
}

File: 20221103_Thursday/20221103.rb
# Leetcode: 2131. Longest Palindrome by Concatenating Two Letter Words.
# https://leetcode.com/problems/longest-palindrome-by-concatenating-two-letter-words/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 2165 ms, faster than 5.26% of Ruby online submissions for Longest Palindrome by Concatenating Two Letter Words.
# Memory Usage: 221.7 MB, less than 73.68% of Ruby online submissions for Longest Palindrome by Concatenating Two Letter Words.
# @param {String[]} words
# @return {Integer}
def longest_palindrome(words)
  duplicates_hash = {}
  # Pairs hash. 
  # Key: Pair sorted ascending.
  # Value: Two values array first Ascending second Descending.
  pairs_hash = {}
  words.each do |w|
    if w[0] == w[1]
      duplicates_hash[w] ||= 0
      duplicates_hash[w] += 1
    elsif w[0] < w[1]
      key = w
      pairs_hash[key] ||= [0,0]
      pairs_hash[key][0] += 1
    elsif w[0] > w[1]
      key = w.reverse
      pairs_hash[key] ||= [0,0]
      pairs_hash[key][1] += 1
    end
  end
  answer = 0
  pairs_hash.keys.each do |key|
    answer += pairs_hash[key].min * 4 # Fixed. Was 2.
  end
  duplicates_hash.keys.each do |key|
    answer += (duplicates_hash[key]/2) * 4
    duplicates_hash[key] %= 2
  end
  duplicates_hash.keys.each do |key|
    if duplicates_hash[key] > 0
      answer += 2
      break
    end
  end
  return answer
end
